### PR TITLE
BasicAuth support In Marathon

### PR DESCRIPTION
**Pull request by gondor**
@beorn7 @juliusv @fabxc 

This is a simple enhancement to support BasicAuth in Marathon.  I also added this to the good config as a part of the test suite.

Our Marathon cluster used LDAP authentication frontend by BasicAuth which also allowed me to test at runtime.

---

This change is [Reviewable](https://reviewable.io/reviews/prometheus/prometheus/1745)

### CODE REVIEW

**Comment by beorn7** [comment]
@xperimental is our Marathon expert. Robert, would you be able to have a look at this?

File: retrieval/discovery/marathon/marathon.go [hunk 1]
```diff
@@ -146,11 +147,22 @@ type AppList struct {
...
+	if err != nil {
+		return nil, err
+	}
+
+	if auth != nil {
+		request.SetBasicAuth(auth.Username, auth.Password)
+	}
+
+	resp, err := http.DefaultClient.Do(request)
+
```
Comment (fabxc) [inline retrieval/discovery/marathon/marathon.go]: Those newlines are a bit excessive.

File: retrieval/discovery/marathon/marathon_test.go [hunk 2]
```diff
@@ -41,7 +41,7 @@ func TestMarathonSDHandleError(t *testing.T) {
 	var (
 		errTesting = errors.New("testing failure")
 		ch         = make(chan []*config.TargetGroup, 1)
-		client     = func(url string) (*AppList, error) { return nil, errTesting }
+		client     = func(url string, auth *config.BasicAuth) (*AppList, error) { return nil, errTesting }
```
Comment (fabxc) [inline retrieval/discovery/marathon/marathon_test.go]: What's up with these method signature extensions? They are not used.
That said, the general feature would need a test.

**Comment by fabxc** [comment]
Thanks.

I assume you could just as well use bearer tokens etc. – should we unify the auth approaches we want to support (i.e. same ones as for scraping) into an object we just provide with every network-based SD?

@brian-brazil

**Comment by brian-brazil** [comment]
> should we unify the auth approaches we want to support (i.e. same ones as for scraping) into an object we just provide with every network-based SD?

Yes, we need it in the AM and blackbox exporter too so this should ultimately go into common.

**Comment by xperimental** [comment]
Again, sorry for the delay. Tested it with a local Marathon installation, the implementation here works fine.

That said, there is an alternative already, because the credentials could always be provided using the URL like this (the disadvantage being, that the password is clearly readable on the configuration page):

```
scrape_configs:
- job_name: local-marathon
  marathon_sd_configs:
  - servers:
    - http://user:password@marathon.server.local:8080
```

Currently the open-source version of Marathon only supports Basic-Authentication, so it would probably not benefit from a "generic" implementation. I can not speak for the commercial version though.

I also noticed that the error message is says pretty much nothing about the actual problem when you have a Marathon server with authentication running and prometheus not configured with the correct credentials:

```
ERRO[0005] Error while updating services: invalid character '<' looking for beginning of value  source=marathon.go:67
```

This has already been the case before, so it would probably be good to fix this even if this PR does not make it. A simple check for the HTTP status code should be enough.

**Comment by ChillkroeteTTS** [comment]
Are there any plans to merge this finally?
Having basic auth for SD with marathon would be great.
Having my credentials in the server url and thus readable for everyone with access to prometheus is not really an option.

**Comment by xperimental** [comment]
@ChillkroeteTTS To be honest I lost track of this issue and don't know what the decision on "should this be merged" was, if there was any.

By now the PR is also a bit outdated and needs rebasing. I had hoped that there was a way to specify the credentials by now when I read your comment the first time, but this sadly does not seem to be the case, so maybe this PR should be revived.

**Comment by grobie** [comment]
@ChillkroeteTTS @xperimental @gondor I'm happy to review an updated PR.

**Comment by brian-brazil** [comment]
Using HTTPClientConfig would be better than just doing basic auth.

**Comment by brian-brazil** [comment]
This still needs to be updated to use our usual http client config code. Feel free to send a new PR with these changes, this one is rather stale by now.

**Pull request closed without merging**